### ISSUE TITLE
PermissionError within subprocess.Popen call

### ISSUE THREAD

**Issue by XCanG** [spyoungtech/ahk#31]
python runned from cmd with admin privileges.
running any action, e.g. `ahk.mouse_move(x=100, y=100, speed=30, blocking=False)`
will throw PermissionError:
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python\Python37\lib\site-packages\ahk\mouse.py", line 89, in mouse_move
    self.run_script(script, blocking=blocking)
  File "C:\Python\Python37\lib\site-packages\ahk\script.py", line 76, in run_script
    result = self._run_script(script_text, decode=decode, blocking=blocking, **runkwargs)
  File "C:\Python\Python37\lib\site-packages\ahk\script.py", line 66, in _run_script
    proc = subprocess.Popen(runargs, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs)
  File "C:\Python\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Python\Python37\lib\subprocess.py", line 1178, in _execute_child
    startupinfo)
PermissionError: [WinError 5] Access is denied
```

Checking permissions from PowerShell:
```powershell
PS D:\Prog\AutoHotkey> Get-Acl | fl


Path   : Microsoft.PowerShell.Core\FileSystem::D:\Prog\AutoHotkey
Owner  : BUILTIN\Administrators
Group  : G:S-1-5-21-1413140394-118887914-2539411025-513
Access : Все Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
         S-1-5-21-1413140394-118887914-2539411025-1000 Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:S-1-5-21-1413140394-118887914-2539411025-513D:AI(A;OICIID;FA;;;WD)(A;OICIID;FA;;;SY)(A;OICIID;FA;;;S-1-5
         -21-1413140394-118887914-2539411025-1000)(A;OICIID;FA;;;BA)



PS D:\Prog\AutoHotkey> Dir | Get-Acl | fl
...
Path   : Microsoft.PowerShell.Core\FileSystem::D:\Prog\AutoHotkey\AutoHotkey.exe
Owner  : BUILTIN\Administrators
Group  : G:S-1-5-21-1413140394-118887914-2539411025-513
Access : Все Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
         S-1-5-21-1413140394-118887914-2539411025-1000 Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:S-1-5-21-1413140394-118887914-2539411025-513D:AI(A;ID;FA;;;WD)(A;ID;FA;;;SY)(A;ID;FA;;;S-1-5-21-14131403
         94-118887914-2539411025-1000)(A;ID;FA;;;BA)

Path   : Microsoft.PowerShell.Core\FileSystem::D:\Prog\AutoHotkey\AutoHotkeyA32.exe
Owner  : BUILTIN\Administrators
Group  : G:S-1-5-21-1413140394-118887914-2539411025-513
Access : Все Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
         S-1-5-21-1413140394-118887914-2539411025-1000 Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:S-1-5-21-1413140394-118887914-2539411025-513D:AI(A;ID;FA;;;WD)(A;ID;FA;;;SY)(A;ID;FA;;;S-1-5-21-14131403
         94-118887914-2539411025-1000)(A;ID;FA;;;BA)

Path   : Microsoft.PowerShell.Core\FileSystem::D:\Prog\AutoHotkey\AutoHotkeyU32.exe
Owner  : BUILTIN\Administrators
Group  : G:S-1-5-21-1413140394-118887914-2539411025-513
Access : Все Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
         S-1-5-21-1413140394-118887914-2539411025-1000 Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:S-1-5-21-1413140394-118887914-2539411025-513D:AI(A;ID;FA;;;WD)(A;ID;FA;;;SY)(A;ID;FA;;;S-1-5-21-14131403
         94-118887914-2539411025-1000)(A;ID;FA;;;BA)

Path   : Microsoft.PowerShell.Core\FileSystem::D:\Prog\AutoHotkey\AutoHotkeyU64.exe
Owner  : BUILTIN\Administrators
Group  : G:S-1-5-21-1413140394-118887914-2539411025-513
Access : Все Allow  FullControl
         NT AUTHORITY\SYSTEM Allow  FullControl
         S-1-5-21-1413140394-118887914-2539411025-1000 Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  :
Sddl   : O:BAG:S-1-5-21-1413140394-118887914-2539411025-513D:AI(A;ID;FA;;;WD)(A;ID;FA;;;SY)(A;ID;FA;;;S-1-5-21-14131403
         94-118887914-2539411025-1000)(A;ID;FA;;;BA)
...
```

It's looks like need to run subprocess from admin privileges as well.
For this need to add to subprosess.Popen these args before running AutoHotkey.exe executable:
```
'runas', '/noprofile', '/user:Administrator',
```

**Comment by spyoungtech** [comment]
This is interesting. Thanks for providing the report and permission details @XCanG 

I haven't been able to reproduce this on my local system running as admin or otherwise in either cmd or powershell. Our CI tests in AppVeyor also run as an administrator.

One possible key difference is that I usually install ahk and place the autohotkey executable in a virtualenv (e.g. in `venv/Scripts`) and let `ahk` discover the executable this way.

I'm still searching for answers on why this permission issue may happen. One possible reason would be a bug in how `executable_path` is being formed. For example, if somehow `executable_path` is a *directory* rather than the actual .exe, it would produce this Permission Denied error. There may be a bug there.  If you are using the `AHK_PATH` env variable or specifying the path by keyword, be sure to use absolute path directly to the executable.

Could you tell me what the output is of 

```
print(ahk.executable_path)
```

Also, try the following steps to setup an environment and see if you have different results....

```
mkdir test
cd test
py -3.7 -m venv venv
venv\Scripts\activate
python -m pip install ahk
copy D:\Prog\AutoHotkey\AutoHotkey.exe venv\Scripts
```
Then try running your script.

Also it would be good to know what version of Windows you are running and if you are in a domain environment or not. And, if possible, provide full code you have for instantiating `ahk`.

**Comment by XCanG** [comment]
```py
>>> print(ahk.executable_path)
D:\Prog\AutoHotkey\
```
----
Running script from virtulenv have same error.

Full code is test code from example:
```py
from ahk import AHK
ahk = AHK()
ahk.mouse_position = (200, 200)
```
last line throw error...

Windows 10 x64 version 10.0.17134

**Comment by spyoungtech** [comment]
Ok, this is interesting. 

So, somehow, the lib is incorrectly detecting the `executable_path`. 

`D:\Prog\AutoHotkey\` should be `D:\Prog\AutoHotkey\AutoHotkey.exe` !


If you do the following, it should alleviate the problem.

```python
ahk = AHK(executable_path='D:\\Prog\\AutoHotkey\\AutoHotkey.exe')
```

Let me know if that fixes things.

Meantime, I'll investigate why this is happening. Basically, if you set `AHK_PATH`, that is used... otherwise `shutil.which` is used to find the executable, which *should* be giving the full path, not just the directory.

**Comment by XCanG** [comment]
Oh, yes, defining path to `.exe` helps, now script run without error.

**Comment by XCanG** [comment]
I guess if you want to use `AHK_PATH` you make little check if it end to `.exe` otherwise you add `os.path.join(AHK_PATH, "AutoHotkey.exe")`.

**Comment by spyoungtech** [comment]
That's definitely a possible solution. 

Technically, the executable can be named anything and the extension is not necessary or could even be something else entirely. For example, if someone really wanted the executable to be named `C:\corp_auto_tool.bin` and have that in `AHK_PATH` and expect it to work.


I think what I'd lean towards in a code-based approach to this would be to add a check and raise an error if this variable is mis-configured.
maybe something like

```
if os.path.isdir(executable_path):
    raise ValueError('AHK_PATH must be a file, not a directory')
```


Anyhow, I'll have to give this some more thought and will probably include a change in the next release. 

Meanwhile, I'll update the docs, too.

If you'd like to make a PR against this, I'd happily review.

**Comment by spyoungtech** [comment]
Marking this one as resolved and using #32 to track progress on action items for changes.

Feel free to reopen or pick up in 32 if you have more questions on this.

Thanks again for the report!
